Stream reuse

wcawijngaards opened the pull request:
This implements stream reuse towards Unbound's upstream servers.  When a query is sent to a server for which there is an open socket, the socket is reused without opening a new stream to the server.  For TCP and TLS queries, for IPv4 and IPv6.  Sockets are opened per-thread, since the per-thread worker keeps track of the reused sockets.

wcawijngaards committed:
Stream reuse branch, for TCP and TLS stream reuse.
This is for upstream pipes and using them again for the next query.

Signposted code for reuse_tcp structure in outside_network.h

wcawijngaards committed:
Merge branch 'master' into stream-reuse

wcawijngaards committed:
Merge branch 'master' into stream-reuse

wcawijngaards committed:
Merge branch 'master' into stream-reuse

wcawijngaards committed:
Merge branch 'master' into stream-reuse

wcawijngaards committed:
Merge branch 'master' into stream-reuse

wcawijngaards committed:
Merge branch 'stream-reuse' of github.com:NLnetLabs/unbound into stream-reuse

wcawijngaards committed:
Merge branch 'master' into stream-reuse

wcawijngaards committed:
Merge branch 'master' into stream-reuse

wcawijngaards committed:
fixup lru list presence boolean.

wcawijngaards committed:
rename next,prev to lru_next,lru_prev for clarity.

wcawijngaards committed:
tree key addr

wcawijngaards committed:
add debug printout

wcawijngaards committed:
add test for tcp reuse

wcawijngaards committed:
find reuse find tcp loop code for multiple connections to the same destination
find spare id value in reused connection.

wcawijngaards committed:
spare id random selection better.

wcawijngaards committed:
fix spare id random selection.

wcawijngaards committed:
tcp connection is stored and picked up for reuse
fix that comm_point_start_listening does not close the same fd that is started.

wcawijngaards committed:
fix to set pending pointer in reuse tcp structure
fix debug output of reuse tcp and test leak of process

wcawijngaards committed:
set timeout to wait for reuse
add comm_point indicator for write events for reuse stream writes.

wcawijngaards committed:
add bool if on tcp waiting list, so that pkt can be stored.
remove pkt args from outnet_tcp_take_into_use, use w.pkt.

wcawijngaards committed:
fix documentation for waiting_tcp pkt NULL setting.

wcawijngaards committed:
fix uninit after malloc for on_tcp_waiting_list.

wcawijngaards committed:
fix crash on cleanup.

wcawijngaards committed:
tcp reuse timeout event cleanup and callbacks.

wcawijngaards committed:
Fix grammar.

wcawijngaards committed:
fix reuse tcp crash, use addr in reuse struct, free leaked tcp entries.

wcawijngaards committed:
reuse tcp lookup with correct address as key.

wcawijngaards committed:
Merge branch 'master' into stream-reuse

wcawijngaards committed:
Merge branch 'master' into stream-reuse

wcawijngaards committed:
reuse tcp id_cmp function. clear list and tree after delete. clear when
decommisioned. callbacks when closed for lru space, and when closed because
not kept open.

wcawijngaards committed:
fix testcode for added function whitelist item.

wcawijngaards committed:
comm point write and read structure members.

wcawijngaards committed:
tcp callback handle timeout event for read and reuse keepalive.

wcawijngaards committed:
in outside_network.c: also log messages that end up on the waiting list.
with dnstap.
for tcp use_free_buffer reuse existing entry if second wait entry on the
same addr as the other waiting.

wcawijngaards committed:
for tcp use_free_buffer write straight away on reuse connection, if possible

wcawijngaards committed:
pending_tcp_query: cleaner comments.

wcawijngaards committed:
outnet_tcp_cb: add assertion and return when write packets done is handled.

wcawijngaards committed:
tcp callback function refactor, split read and timeout event setup, leave
unused queries that are already sent to track their reply on the query
pipeline, when serviced query is deleted deal with the write wait list,

wcawijngaards committed:
in tcp write callback routine dont reset read byte count if write and read.
in tcp write callback and write and read and write is done perform callback.
tcp connection is selected to not toggle readwrite and give closed callback.

wcawijngaards committed:
tcp read and write handling of write events in netevent for tcp and ssl.

wcawijngaards committed:
Merge branch 'master' into stream-reuse

wcawijngaards committed:
in outnet_tcptimer: pick up callbacks and clean the struct pending for
use again in tcp_free list; then perform the callbacks in a cleaner state
for reentry into the outside network code.  Delete callbacks afterwards.

wcawijngaards committed:
stream reuse toggle write and read to only read mode when write is done.
tcp callback from stream read without entry linked.

wcawijngaards committed:
debug prints in verbose output.

wcawijngaards committed:
stream reuse, check incoming messages from rbtree in outnet_tcp_cb when
read a reply, and call that callback and remove that entry.

wcawijngaards committed:
stream reuse, the id for pending stored in waiting_tcp structure, because
there can be multiple messages waiting for reply.

wcawijngaards committed:
stream reuse, fix to return key pointer from reuse_tcp_by_id_find.

wcawijngaards committed:
stream reuse, fix to not keep stream when it is in error and closed.
stream reuse, fix to stop listening on the fd before rw mode change.
stream reuse, fix waiting tcp id value stored before tree insert.
stream reuse, fix to not double delete pend.query.

wcawijngaards committed:
stream reuse, fix to put id number in waiting tcp packet.

wcawijngaards committed:
stream reuse, fix double callback and double delete, items are in the tree.

wcawijngaards committed:
stream reuse, fix decommission to first remove from tree and then do the
callbacks and then delete the element.

wcawijngaards committed:
stream reuse, comment improved

wcawijngaards committed:
stream reuse, neater code for tree by id and use callback routine for timer.

wcawijngaards committed:
stream reuse, fix bad id in reply errors.

wcawijngaards committed:
fix that ssl_handle_it() uses tcp_is_reading in tcp_write_and_read mode.
fix that netevent tcp_write_and_read mode does not close due to error from
assert that not both write and read happen at the same time.

wcawijngaards committed:
stream reuse, remove debug output

wcawijngaards committed:
stream reuse, write and read again if more data can go over the channel,
this amortizes the event loop mechanism for busy channels, for performance.

wcawijngaards committed:
stream reuse, in callbacks, removed whitespace.

wcawijngaards committed:
stream reuse, update lru when reuse elements are used with lru touch routine.

wcawijngaards committed:
stream reuse, make reuse possible straight away after first query to address.

wcawijngaards committed:
Merge branch 'master' into stream-reuse

wcawijngaards committed:
Merge branch 'master' into stream-reuse

wcawijngaards committed:
Merge branch 'master' into stream-reuse

wcawijngaards committed:
stream reuse, free up elements that are connected in outside network delete.

wcawijngaards committed:
stream reuse, fix cleanup with streams in the connection table.

wcawijngaards committed:
stream reuse, test timeout and simultaneous queries.

wcawijngaards committed:
stream reuse, test with a list of outstanding queries to the upstream server.

wcawijngaards committed:
stream reuse, test with connection drops.

wcawijngaards committed:
Merge branch 'master' into stream-reuse

wcawijngaards committed:
stream reuse, send queries one by one when upstream refuses multiple queries,
by closing the connection.

wcawijngaards committed:
stream reuse, defensible wait add and debug log with details for requeue.

wcawijngaards committed:
stream reuse, add a test for close by upstream server after timeout.

wcawijngaards committed:
stream reuse, fix tls close by upstream after timeout write event handling.

wcawijngaards committed:
stream reuse, add tls test for stream reuse.

wcawijngaards committed:
stream reuse, disable debug in test

wcawijngaards committed:
Merge branch 'master' into stream-reuse

wcawijngaards committed:
stream reuse, move drop in tcp_reuse test to timeout section of test.

wcawijngaards committed:
Merge branch 'master' into stream-reuse

wcawijngaards commented:
This PR fixes issue #47

wcawijngaards committed:
Review fix: remove unused variables.

wcawijngaards committed:
Review fix: defense check of qdcount in debug output.

pyropeter commented:
Will this be part of 1.12.0 or are you in feature freeze?

wcawijngaards commented:
This is not in 1.12.0, because we are in feature freeze for it.  It is waiting for review, and maybe some fixes after that.

djcater commented:
Will this cover both upstream DNS-Over-TLS connections and upstream DNS-Over-HTTPS connections?

ralphdolmans commented:
> Will this cover both upstream DNS-Over-TLS connections and upstream DNS-Over-HTTPS connections?

There is no upstream DoH support in Unbound yet, only downstream. If upstream DoH will be added at a later stage it will benefit from the changes in the PR.

ralphdolmans reviewed (approved):
Looks good to me!

wcawijngaards committed:
- Fix that reuse_tcp_close_oldest sets item_on_lru_list to 0.
- Fix to add assertions to reuse_tcp_select_id and unit test.
- Fix that if no tcp buffers then pending tcp query stops.

wcawijngaards committed:
Merge branch 'master' into stream-reuse
Resolved comm_point_start_listening conflict for socket close.

wcawijngaards committed:
Merge branch 'master' into stream-reuse

iz8mbw commented:
Ready to merge? 😁

wcawijngaards commented:
The branch is ready for use, it is waiting for review(s), before merge.

gthess commented on services/outside_network.c:
```diff
@@ -383,50 +648,244 @@ outnet_tcp_take_into_use(struct waiting_tcp* w, uint8_t* pkt, size_t pkt_len)
+		pend->c->tcp_write_pkt_len == pend->query->pkt_len) {
+		/* since the current query is not written, it can also
+		 * move to a free buffer */
+		if(verbosity >= 5 && pend->query->pkt_len > 12+2+2 &&
```
Verbosity number check

wcawijngaards replied:
fixed.

gthess commented on services/outside_network.c:
```diff
@@ -383,50 +648,244 @@ outnet_tcp_take_into_use(struct waiting_tcp* w, uint8_t* pkt, size_t pkt_len)
+		outnet_add_tcp_waiting(outnet, w);
+	}
+	while((w = reuse_write_wait_pop(&pend->reuse)) != NULL) {
+		if(verbosity >= 5 && w->pkt_len > 12+2+2 &&
```
Verbosity number check

wcawijngaards replied:
fixed.

gthess commented on services/outside_network.c:
```diff
@@ -437,34 +896,184 @@ decommission_pending_tcp(struct outside_network* outnet,
+		/* if we are writing, ignore readtimer, wait for write timer
+		 * or write is done */
+		if(pend->c->tcp_write_and_read)
+			return 0;
```
The two `if(pend->c->tcp_write_and_read)` can be combined

wcawijngaards replied:
okay.

gthess commented on services/outside_network.c:
```diff
@@ -1273,93 +1899,229 @@ outnet_tcptimer(void* arg)
+	log_assert(reuse->tree_by_id.count < 0xffff);
+	select = ub_random_max(outnet->rnd, 0xffff - reuse->tree_by_id.count);
+	/* select value now in 0 .. num free - 1 */
+
+	count = 0; /* number of free spaces passed by */
+	node = rbtree_first(&reuse->tree_by_id);
+	log_assert(node && node != RBTREE_NULL); /* tree not empty */
+	/* see if select is before first node */
+	if(select < tree_by_id_get_id(node))
+		return select;
+	count += tree_by_id_get_id(node);
+	/* perhaps select is between nodes */
+	while(node && node != RBTREE_NULL) {
+		rbnode_type* next = rbtree_next(node);
+		if(next && next != RBTREE_NULL) {
+			curid = tree_by_id_get_id(node);
+			nextid = tree_by_id_get_id(next);
+			log_assert(curid < nextid);
+			if(curid != 0xffff && curid + 1 < nextid) {
+				/* space between nodes */
+				space = nextid - curid - 1;
+				log_assert(select >= count);
+				if(select < count + space) {
+					/* here it is */
+					return curid + 1 + (select - count);
+				}
+				count += space;
+			}
+		}
+		node = next;
+	}
+
+	/* select is after the last node */
+	/* count is the number of free positions before the nodes in the
+	 * tree */
+	node = rbtree_last(&reuse->tree_by_id);
+	log_assert(node && node != RBTREE_NULL); /* tree not empty */
+	curid = tree_by_id_get_id(node);
+	log_assert(count + (0xffff-curid) + reuse->tree_by_id.count == 0xffff);
+	return curid + 1 + (select - count);
```
Just a note: this last resort seems less random to me but since we only need the ID for differentiating the queries in the same TCP connection I don't see a problem.

wcawijngaards replied:
It is the same randomness? I thought?  Because it picks a random item from the list of available items?

gthess replied:
Upon closer reading this looks good!

gthess commented on services/outside_network.h:
```diff
@@ -211,6 +227,59 @@ struct port_comm {
+	/** rbtree node with links in tcp_reuse tree. key is NULL when not
+	 * in tree. Both active and empty connections are in the tree.
+	 * key is this structure, the sockaddr and and then is-ssl bool,
+	 * and then ptr value for several times same address in tree */
```
The comment about the key is not very clear.

wcawijngaards replied:
Updated it.

gthess commented on testdata/ssl_reuse.tdir/ssl_reuse.conf:
```diff
@@ -0,0 +1,18 @@
+	do-not-query-localhost: no
+
+	tls-cert-bundle: "unbound_server.pem"
+	ssl-upstream: yes
```
Could use the `tls-upstream` option instead.

wcawijngaards replied:
fixed.

gthess commented on testdata/ssl_reuse.tdir/ssl_reuse.test:
```diff
@@ -0,0 +1,308 @@
+
+# make the server timeout to drop the upstream connection
+echo "> sleep 20"
+sleep 15
```
Mismatch between print value and command.

wcawijngaards replied:
fixed.

gthess commented on testdata/tcp_reuse.tdir/tcp_reuse.test:
```diff
@@ -0,0 +1,309 @@
+
+# make the server timeout to drop the upstream connection
+echo "> sleep 20"
+sleep 15
```
Mismatch between print value and command.

wcawijngaards replied:
fixed.

gthess reviewed (changes requested):
Looks good to me also!

I have a question and some nits.
What is the benefit of `MAX_REUSE_TCP_QUERIES`?

The nits:
1. Shouldn't `log_reuse_tcp(<number>, ...)` turn to `log_reuse_tcp(<level>, ...)` and
  `verbose(<number>, ...)` turn to `verbose(<level>, ...)`? Unless they are for dev debugging and will be removed?
2. The `ssl_reuse.tdir` and the `ssl_reuse.*` files therein could be renamed to tls_reuse.

wcawijngaards commented:
The `MAX_REUSE_TCP_QUERIES` is so that it does not make too many queries at the same time to a destination.  This is because it starts to queue up, but then in terms of TCP buffers, and also, if the connection fails for some reason that is a lot of queries.  So I thought it best not to have too many at the same time on one stream, at the level of thousands it should remove overhead costs just fine.

iz8mbw commented:
Hi. A my personal note, if I have correctly understood, the connection timeout of reuse is hard-coded to 30 seconds.
This means that the ESTHABILISHED TCP connection to forwarder remains ESTHABILISHED up to 30 seconds (if no new DNS query are made within these 30 seconds).
I think 30 seconds is a low value because if no DNS queries are made for 30 seconds the reuse is useless because unbound will go to create new connections to forwarder.
Can be this value increased? I think 60 seconds make more sense.
Thank you.

wcawijngaards committed:
stream reuse, debug output with verbose level instead of number.

wcawijngaards committed:
stream reuse, renamed ssl_reuse.tdir to tls_reuse.tdir.

wcawijngaards committed:
stream reuse, rename ssl_reuse test to tls_reuse test.

wcawijngaards committed:
stream reuse, fix review comments.

wcawijngaards committed:
stream reuse, review comments.

wcawijngaards committed:
stream reuse, fix review comments.

wcawijngaards committed:
stream reuse, fix review comments.

wcawijngaards committed:
Merge branch 'master' into stream-reuse

wcawijngaards committed:
stream reuse, up connection reuse time to 60 seconds.

wcawijngaards commented:
Sure, increased the reuse time to 60 seconds in https://github.com/NLnetLabs/unbound/pull/283/commits/6b97cb1dd127125f99ac44df62bfcfc07b5fe1c3 .

gthess reviewed (approved):
Thanks for incorporating the feedback, looks good to me!

wcawijngaards committed:
Merge branch 'master' into stream-reuse

pengelana commented:
Unbound 1.12.0 with patch from #283 and `--enable-tfo-client` & `--enable-tfo-server` will produce error:

`error: read (in tcp s): Transport endpoint is not connected for xxx.xxx.xxx.xxx port 53`

~~~
dig @127.0.0.1 fedoraproject.org

;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 40656
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;fedoraproject.org.             IN      A

;; Query time: 213 msec
~~~

wcawijngaards commented:
The fix is in commit https://github.com/NLnetLabs/unbound/commit/0502ab3026af59673b21133eb7c5a72126b37d66 . Thanks for the report!
